File: doc/resources/CaseChangingCharStream.java
package org.antlr.v4.runtime;

import org.antlr.v4.runtime.misc.Interval;

/**
 * This class supports case-insensitive lexing by wrapping an existing
 * {@link CharStream} and forcing the lexer to see either upper or
 * lowercase characters. Grammar literals should then be either upper or
 * lower case such as 'BEGIN' or 'begin'. The text of the character
 * stream is unaffected. Example: input 'BeGiN' would match lexer rule
 * 'BEGIN' if constructor parameter upper=true but getText() would return
 * 'BeGiN'.
 */
public class CaseChangingCharStream implements CharStream {

	final CharStream stream;
	final boolean upper;

	/**
	 * Constructs a new CaseChangingCharStream wrapping the given {@link CharStream} forcing
	 * all characters to upper case or lower case.
	 * @param stream The stream to wrap.
	 * @param upper If true force each symbol to upper case, otherwise force to lower.
	 */
	public CaseChangingCharStream(CharStream stream, boolean upper) {
		this.stream = stream;
		this.upper = upper;
	}

	@Override
	public String getText(Interval interval) {
		return stream.getText(interval);
	}

	@Override
	public void consume() {
		stream.consume();
	}

	@Override
	public int LA(int i) {
		int c = stream.LA(i);
		if (c <= 0) {
			return c;
		}
		if (upper) {
			return Character.toUpperCase(c);
		}
		return Character.toLowerCase(c);
	}

	@Override
	public int mark() {
		return stream.mark();
	}

	@Override
	public void release(int marker) {
		stream.release(marker);
	}

	@Override
	public int index() {
		return stream.index();
	}

	@Override
	public void seek(int index) {
		stream.seek(index);
	}

	@Override
	public int size() {
		return stream.size();
	}

	@Override
	public String getSourceName() {
		return stream.getSourceName();
	}
}

File: doc/resources/case_changing_char_stream.dart
// @dart=2.12

import '../../runtime/Dart/lib/antlr4.dart';
import '../../runtime/Dart/lib/src/interval_set.dart';

/// This class supports case-insensitive lexing by wrapping an existing
/// {@link CharStream} and forcing the lexer to see either upper or
/// lowercase characters. Grammar literals should then be either upper or
/// lower case such as 'BEGIN' or 'begin'. The text of the character
/// stream is unaffected. Example: input 'BeGiN' would match lexer rule
/// 'BEGIN' if constructor parameter upper=true but getText() would return
/// 'BeGiN'.
class CaseChangingCharStream extends CharStream {
  final CharStream stream;
  final bool upper;

  /// Constructs a new CaseChangingCharStream wrapping the given [stream] forcing
  /// all characters to upper case or lower case depending on [upper].
  CaseChangingCharStream(this.stream, this.upper);

  @override
  int? LA(int i) {
    int? c = stream.LA(i);
    if (c == null || c <= 0) {
      return c;
    }
    String newCaseStr;
    if (upper) {
      newCaseStr = String.fromCharCode(c).toUpperCase();
    } else {
      newCaseStr = String.fromCharCode(c).toLowerCase();
    }
    // Skip changing case if length changes (e.g., ß -> SS).
    if (newCaseStr.length != 1) {
      return c;
    } else {
      return newCaseStr.codeUnitAt(0);
    }
  }

  @override
  String get sourceName => stream.sourceName;

  @override
  void consume() => stream.consume();

  @override
  String getText(Interval interval) => stream.getText(interval);

  @override
  int get index => stream.index;

  @override
  int mark() => stream.mark();

  @override
  void release(int marker) => stream.release(marker);

  @override
  void seek(int index) => stream.seek(index);

  @override
  int get size => stream.size;
}
